tpu-client/src/nonblocking/tpu_client.rs

```diff
@@ -478,45 +554,70 @@ where
             }
 
             let mut last_resend = Instant::now() - TRANSACTION_RESEND_INTERVAL;
-            while block_height <= last_valid_block_height {
+            while progress.block_height <= progress.last_valid_block_height {
                 let num_transactions = pending_transactions.len();
 
                 // Periodically re-send all pending transactions
                 if Instant::now().duration_since(last_resend) > TRANSACTION_RESEND_INTERVAL {
+                    // Prepare futures for all transactions
+                    let mut futures = vec![];
                     for (index, (_i, transaction)) in pending_transactions.values().enumerate() {
-                        if !self.send_transaction(transaction).await {
+                        let wire_transaction = serialize(transaction).unwrap();
+                        let leaders = self
+                            .leader_tpu_service
+                            .leader_tpu_sockets(self.fanout_slots);
+                        futures.extend(send_wire_transaction_futures(
+                            &progress_bar,
+                            &progress,
+                            index,
+                            num_transactions,
+                            wire_transaction,
+                            leaders,
+                            &self.connection_cache,
+                        ));
+                    }
+
+                    // Send them all at once
+                    let results = join_all(futures).await;
```

Comment:
can we transpose all those timeout wrappers by just timeout-wrapping the `JoinAll` future here, or do we want the granularity?

Reply:
We want the granularity unfortunately.

We're mimicking the old behavior by waiting 10ms between each send, and that timeout encompasses the waiting too. So for example, if we set a global timeout of 5 seconds, and we want to send 1000 transactions, those last 500 transactions will never get sent since they'll timeout just waiting their turn.